pkg/operators/wasm/datasource.go

```diff
@@ -0,0 +1,525 @@
+		stack[0] = 0
+		return
+	}
+
+	ds, err := i.gadgetCtx.RegisterDataSource(datasource.Type(dsType), dsName)
+	if err != nil {
+		i.logger.Warnf("failed to register datasource: %v", err)
+		stack[0] = 0
+		return
+	}
+	stack[0] = wapi.EncodeU32(i.addHandle(ds))
+}
+
+// getDataSource returns a data source by its name.
+// Params:
+// - stack[0] is the name of the datasource (string encoded)
+// Return value:
+// - DataSource handle on success, 0 on error
+func (i *wasmOperatorInstance) getDataSource(ctx context.Context, m wapi.Module, stack []uint64) {
+	dsNamePtr := stack[0]
+
+	dsName, err := stringFromStack(m, dsNamePtr)
+	if err != nil {
+		i.logger.Warnf("getDataSource: reading string from stack: %v", err)
+		stack[0] = 0
+		return
+	}
+	ds := i.gadgetCtx.GetDataSources()[dsName]
+	if ds == nil {
+		i.logger.Warnf("datasource not found %q", dsName)
+		stack[0] = 0
+		return
+	}
+	stack[0] = wapi.EncodeU32(i.addHandle(ds))
+}
+
+// dataSourceGetField returns a handle to a data source.
+// Params:
+// - stack[0]: DataSource handle
+// - stack[1]: Field name
+// Return value:
+// - Field handle on success, 0 on error
+func (i *wasmOperatorInstance) dataSourceGetField(ctx context.Context, m wapi.Module, stack []uint64) {
+	dsHandle := wapi.DecodeU32(stack[0])
+	fieldNamePtr := stack[1]
+
+	ds, ok := getHandle[datasource.DataSource](i, dsHandle)
+	if !ok {
+		stack[0] = 0
+		return
+	}
+	fieldName, err := stringFromStack(m, fieldNamePtr)
+	if err != nil {
+		i.logger.Warnf("dataSourceGetField: reading string from stack: %v", err)
+		stack[0] = 0
+		return
+	}
+	acc := ds.GetField(fieldName)
+	stack[0] = wapi.EncodeU32(i.addHandle(acc))
+}
+
+// dataSourceAddField add a field to the data source and returns its handle.
+// Params:
+// - stack[0]: DataSource handle
+// - stack[1]: Field name
+// - stack[2]: Field kind
+// Return value:
+// - Field handle on success, 0 on error
+func (i *wasmOperatorInstance) dataSourceAddField(ctx context.Context, m wapi.Module, stack []uint64) {
+	dsHandle := wapi.DecodeU32(stack[0])
+	fieldNamePtr := stack[1]
+	fieldKind := wapi.DecodeU32(stack[2])
+	// TODO: add kind max?
+	if fieldKind > uint32(api.Kind_Bytes) {
+		i.logger.Warnf("dataSourceAddField: invalid field kind %d", fieldKind)
+		stack[0] = 0
+		return
+	}
+
+	ds, ok := getHandle[datasource.DataSource](i, dsHandle)
+	if !ok {
+		stack[0] = 0
+		return
+	}
+	fieldName, err := stringFromStack(m, fieldNamePtr)
+	if err != nil {
+		i.logger.Warnf("dataSourceAddField: reading string from stack: %v", err)
+		stack[0] = 0
+		return
+	}
+	acc, err := ds.AddField(fieldName, api.Kind(fieldKind))
+	if err != nil {
+		i.logger.Warnf("adding field %q to datasource %q: %v", fieldName, ds.Name(), err)
+		stack[0] = 0
+		return
+	}
+	stack[0] = wapi.EncodeU32(i.addHandle(acc))
+}
+
+type subscriptionType uint32
+
+const (
+	subscriptionTypeData subscriptionType = iota
+	subscriptionTypeArray
+	subscriptionTypePacket
```

Comment:
For APIs like this, I think it's better to hardcode the values instead of using iota - I've seen that you added checks to keep it in sync, but for external implementations (other languages) it could break stuff if we accidentally just inserted a new type.